lib/value/src/value/iter.rs

```diff
@@ -0,0 +1,466 @@
+use crate::Value;
+use std::{marker::PhantomData, ops::IndexMut};
+
+impl Value {
+    /// Create an iterater over the `Value`.
+    ///
+    /// For non-collection types, this returns a single-item iterator similar to
+    /// `Option`'s iterator implementation.
+    ///
+    /// For collection types, it returns all elements in the collection.
+    ///
+    /// The resulting item is an [`IterItem`], which contains either a mutable
+    /// `Value` for non-collection types, a (`&mut String`, `&mut Value`) pair for
+    /// object-type collections, or an immutable/mutable (`usize`, `&mut Value`)
+    /// pair for array-type collections.
+    ///
+    /// ## Recursion
+    ///
+    /// If `recursion` is set to `true`, the iterator recurses into nested
+    /// collection types.
+    ///
+    /// Recursion follows these rules:
+    ///
+    /// - When a collection type is found, that type is first returned as-is.
+    ///   That is, if we're iterating over an object, and within that object is
+    ///   a field "foo" containing an array, then we first return
+    ///   `IterItem::KeyValue`, which contains the key "foo", and the array as
+    ///   the value.
+    /// - After returning the collection type, the iterator recurses into the
+    ///   nested collection itself. Using the previous example, we now go into
+    ///   the array, and start returning `IterItem::IndexValue` variants for the
+    ///   elements within the array.
+    /// - Any mutations done to the array before recursing into it are
+    ///   preserved, meaning once recursion starts, the mutations done on the
+    ///   object itself are preserved.
+    pub fn into_iter<'a>(self, recursive: bool) -> ValueIter<'a> {
+        let data = match self {
+            Value::Object(object) => IterData::Object(object.into_iter().collect()),
+            Value::Array(array) => IterData::Array(array),
+            value => IterData::Value(value),
+        };
+
+        ValueIter::new(data, recursive)
+    }
+}
+
+/// An [`Iterator`] over a [`Value`].
+pub struct ValueIter<'a> {
+    data: IterData,
+    recursive: bool,
+    index: usize,
+    recursive_iter: Option<Box<Self>>,
+    must_prepare_recursion: bool,
+    phantom: PhantomData<&'a mut ()>,
+}
+
+/// The [`Iterator::Item`] returned by the [`ValueIter`] iterator.
+#[derive(Debug, PartialEq, Eq)]
+#[allow(clippy::enum_variant_names)]
+pub enum IterItem<'a> {
+    /// A single primitive value.
+    Value(&'a mut Value),
+
+    /// A key/value combination.
+    KeyValue(&'a mut String, &'a mut Value),
+
+    /// An index/value combination.
+    IndexValue(usize, &'a mut Value),
+}
+
+/// The internal data representation used by the iterator.
+///
+/// Objects are stored as a `Vec<(String, Value)>`, to allow mutating the object
+/// keys during iteration.
+enum IterData {
+    Value(Value),
+    Object(Vec<(String, Value)>),
+    Array(Vec<Value>),
```

Comment:
If storing the iterator, this becomes something like:
```suggestion
    Object(std::collections::btree_map::IntoIter<String, Value>),
    Array(std::vec::IntoIter<Value>),
```